### DISCUSSION TITLE
EPS to PNG Conversion issues

### DISCUSSION THREAD

**Discussion by hsweet3** [ImageMagick/ImageMagick#5831] [Help] [answered]
Version: ImageMagick 7.1.0-52 Q16-HDRI x64 04ee6ce:20221106 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raqm raw rsvg tiff webp xml zip zlib
Compiler: Visual Studio 2022 (193331630)

>>>>

I am attempting to convert eps to gif (or png?) but my eps is CMYK so the colors don't come out right.
if i attempt to convert to jpg I am getting good color (not perfect) but the resulting file size of the jpg is larger size than input eps file
[783203ng.zip](https://github.com/ImageMagick/ImageMagick/files/10115295/783203ng.zip)
```
identify 783203ng.eps
783203ng.eps[0] EPT **139x162** 139x162+0+0 16-bit ColorSeparation CMYK 90173B 0.094u 0:00.044
783203ng.eps[1] TIFF 139x162 139x162+0+0 8-bit sRGB 256c 48323B 0.047u 0:00.024
```

I've been using "
`magick -density 300 783203ng.eps[0] 783203_ng.jpg`
 to convert but the resulting jpg is larger

```
identify 783203_ng.jpg
783203_ng.jpg JPEG **578x675** 578x675+0+0 8-bit CMYK 185873B 0.000u 0:00.002
```

the -density command seems to create a larger file?
I can create the proper output size in the jpg by indicating `_-resize 139x162_` but I have multiple files with all different sizes to convert. Is there a way to read the geometry of the original file and apply to the resulting jpg using -resize?

thanks in advance

**Comment by snibgo** [comment]
> the -density command seems to create a larger file?

I assume you mean the width and height have more pixels. If you don't want that, then don't use "-density".

**Reply by hsweet3** [reply]
if I don't use -density to render the vector file before conversion the result is grainy and awful... it does convert as the same size but is unusable

**Comment by snibgo** [comment, marked as answer]
You can read it once without "-density" to get the dimensions, then read it again with whatever density you want and resize that to the found dimensions.
```
magick 783203ng.eps[0] -set option:MYSIZE %wx%h +delete -density 600 783203ng.eps[0] -resize %[MYSIZE] out.png
```

**Reply by hsweet3** [reply]
@snibgo your above command takes care of my question of finding the right size dimensions of the original and applying to the output! thank you. Great feedback.

Can I ask a couple more questions?
Since the eps file is created in CMYK and am going to convert to jpg but would rather use gif because they are for web, and because I need to use _-transparent_ for background

so to use gif or png I've tried converting from colorspace cmyk to colorspace srgb, tried different profile statements with .icc profiles but I can't get a color match.

As you see in my 2 identify statements above in the alpha channel of the eps is 16-bit CMYK but the resulting jpg is 8-bit CMYK.
Can I 'force' 16-but CMYK in the resulting jpg?
Is there a comparable command for jpg to get the same result as transparent in gif?

thanks again for the "MYSIZE" that's a big deal for me!

**Reply by hsweet3** [reply]
I did find a command somewhere which handles my colors exactly as I'd hoped but I suspect it's an older version because of the "convert" command
`convert -strip -colorspace sRGB -density 600 -background white -flatten -resize 600 -colors 256 -depth 16 783203ng.eps[0] converted_ng203.gif`



I just don't know how to combine the 2 commands to get the MYSIZE and to get the colors right and to add the -transparent white variable

I appreciate any help

**Comment by snibgo** [comment]
"MYSIZE" is a name for a variable. You can use any name you like, any case (eg "myPreferredSize").

Reading the EPS file rasterizes it (convert it to pixels). In my "magick" command, you can insert "-colorspace sRGB" to make the rasterization result sRGB (plus alpha) instead of CMYK.

The output for me is 16 bits/channel.

> Is there a comparable command for jpg to get the same result as transparent in gif?

JPEG cannot record transparency.

"-background white -flatten" will flatten against a white background. So all fully-transparent pixels will become opaque white.

I am unsure what result you want. If you simply want a GIF with transparency, then:
```
magick -colorspace sRGB 783203ng.eps[0] -set option:MYSIZE %wx%h +delete -density 600 783203ng.eps[0] -resize %[MYSIZE] out.gif
```
The transparent pixels are transparent black. If you want transparent white, then:
```
magick -colorspace sRGB 783203ng.eps[0] -set option:MYSIZE %wx%h +delete -density 600 783203ng.eps[0] -resize %[MYSIZE] -transparent-color White out.gif
```

**Reply by hsweet3** [reply]
Your feedback has been invaluable! 
the goal(s) were to 

1. find the size (geometry) and pass it to the converted file (gif)
2. Get the color scheme right (strip CMYK/convert it to sRGB)
3. Transparent white

You've helped me to figure all of that out. I thank you.

now I just have to do some testing to see if I have to smooth edges in the gif

thanks again!

**Comment by hsweet3** [comment]
one more issue we're seeing where EPS files contain gradients (like the above image - you can see the "stripes" in his shirt and the pot he's holding.
Can you help me use the following command (or something like it)

`magick -density 1200 -units pixelsperinch -colorspace sRGB 783203ng.eps[0] -resize 25% kidng.png`

which results in this image - which is almost identical to the eps file



and be able to apply the original size to it  - with the MYSIZE variable you showed me last time?

my attempt is 
`magick -quiet -colorspace RGB 783203ng.eps[0] -set option:GETSIZE %w%h +delete -density 1200 -units pixelsperinch 783203ng.eps[0] -resize %[GETSIZE] clearkid.png`
**and it errors out**
magick: unable to extend cache '783203ng.eps': No space left on device @ error/cache.c/OpenPixelCache/3910.

I'm running a PC with 16GB RAM - maybe I can run it on a beefier machine if that's right command...
not sure if I can capture the size of the eps file in a separate command and then apply it after I do the pixelsperinch thing?

Hope this is somewhat of a clear question

**Reply by fmw42** [reply]
Change %w%h.  to %wx%h.  Also why are you using -colorspace RGB before reading the eps file rather than -colorspace sRGB.  They are different!

**Comment by snibgo** [comment]
As fmw42 says.

You resize to GETSIZE, which should be two integers separated by "x", size as "139x162". But what do you set GETSIZE to? "%w%h" with no "x", so GETSIZE is "139162". When you supply only one dimension, the other is assumed to be the same. So you are trying to resize to 139162x139162 pixels, which is 19 giga pixels. Oops.

**Reply by hsweet3** [reply]
19 giga pixels - that explains this behavior :)
![image](https://user-images.githubusercontent.com/22641367/205189012-0dfaa8a9-4a4a-4240-88f5-d987baca3078.png)


ah, so you're saying I should have used my good eye? Makes perfect sense to me now that I see it.
so if I add the x and change back to sRGB

`magick -quiet -colorspace RGB 783203ng.eps[0] -set option:GETSIZE %wx%h +delete -density 1200 -units pixelsperinch 783203ng.eps[0] -resize %[GETSIZE] -transparent-color White clearkid_transparent.png`

the resulting file is tiny. it says it's 139x162 if I do identify or if I open it in Photoshop or Illustrator, but in inches it's miniscule.



the reason I'm using the pixelsperinch is because there are gradients - is that making the images smaller?
or maybe they're not smaller at all... it's just the Adobe products opening them and doing some sort of conversion on open?

identify 783203ng.eps
783203ng.eps[0] EPT 139x162 139x162+0+0 16-bit ColorSeparation CMYK 90173B 0.125u 0:00.074
783203ng.eps[1] TIFF 139x162 139x162+0+0 8-bit sRGB 256c 48323B 0.078u 0:00.038

identify clearkid_transparent.png
clearkid_transparent.png PNG 139x162 139x162+0+0 16-bit sRGB 127037B 0.000u 0:00.002

**Comment by snibgo** [comment]
>  it says it's 139x162 if I do identify or if I open it in Photoshop or Illustrator, but in inches it's miniscule.

That is because the density is 1200 dpi. Just before writing the output file you can "-density 300" or whatever you like. That won't change the pixels, just the metadata. The first read of the EPS file hasn't specified the density, so IM has used 72. That would be the technically correct final density.

**Reply by hsweet3** [reply]
Is this what you mean by 'just before the output file?
`magick -quiet -colorspace sRGB 783203ng.eps[0] -set option:GETSIZE %wx%h +delete -density 1200 -units pixelsperinch 783203ng.eps[0] -density 300 -resize %[GETSIZE] -transparent-color White  yokid.png`

I added the "-density 1200" because the gradients in many images render very well

If I open it in Photoshop it indicates the resolution is correct (300 DPI) but somehow the actual dimensions of the image is smaller (in inches) - also shows me same in Illustrator
![image](https://user-images.githubusercontent.com/22641367/205325594-9a5d420e-cd2f-4ede-93e5-8632cd649ec4.png)

Any help you can provide to get the actual inches of the file to match original will be much appreciated

**Reply by hsweet3** [reply]
I think I found what I'm trying to do

`magick -density 1200 -units pixelsperinch -colorspace sRGB 783203ng.eps[0] -resize 25% -density 300 -transparent-color White kiddec2final.png`



I believe where I was tripped up was the difference between the actual image size and the canvas size. I'm not sure.
Identify command only sees the actual image in pixels (EPS)? therefore when I was seeing the correct pixels but the image itself was smaller.

I have 1 clarifying question
if I use the density 1200 to render it and then say -resize 25% will i consistently get 300DPI in the output? I just doing the simple math of 1200/4 = 300.

Also - the output pngs seem to be about twice the file size.
is there a way to reduce the file size using -quality or some other reduction?

**Reply by hsweet3** [reply]
I appreciate your patience. I've been bouncing between the GETSIZE variable and just letting IM do the conversion without specifying what the output pixels should be

if I use
magick -quiet -colorspace sRGB FFO.eps[0] -set option:GETSIZE %wx%h +delete -density 1200 -units pixelsperinch FFO.eps[0] -resize %[GETSIZE] -transparent-color White -density 300 FFO.png 

I get the 300dpi file as output but there's bit of blur in the resulting file (as compared to a PNG created right out of the Adobe product the eps is created.
2 images - each for the 2 files in the .zip I attaced
![FFO](https://user-images.githubusercontent.com/22641367/205420724-bd9df00a-6c06-4569-9b5a-0829afa5e24c.png)
![IWT](https://user-images.githubusercontent.com/22641367/205420744-9354f39a-c1cf-4cc5-bda9-3e64619a6a80.png)

it's a small amount of blur, but enough to be noticable

I tried sharpen and messed with mogrify a bit but to no avail. I have seen you in a couple of threads in the legacy board talking about setting different densities etc but I have not quite nailed down the blur, and unfortunately it's a blocker at this moment. I assume there's a setting I have wrong or have not included but I don't know enough about IM to say.

I appreciate you sharing your time and expertise

[Final.zip](https://github.com/ImageMagick/ImageMagick/files/10145546/Final.zip)

**Reply by fmw42** [reply]
The best way I know is to increase -density 1200 and then resize to your desired pixel dimensions.  So try 2400 for example. But that will be 4x slower (approx).  So increase density as much as needed to get better quality. However, you cannot compare the quality of a vector file to the raster result. Vectors are always sharp.

**Reply by hsweet3** [reply]
Thanks Fred.
I will test with increasing the density when rendering the image. Hadn't thought of that. The theory there is if I render the input at a higher res and then take the snapshot of it at -density 300 it could be a better result?
Your last comment had me thinking that the advantage of vector is because it's points and lines I can blow it up to the size of a house and there's no image loss/blur - so really if I increase -density on render it's not going to matter... does that sound right?

maybe if when I'm outputting I say density 600 and then resize 50%? or am I limited to 1 resize in a command?

thanks

**Reply by fmw42** [reply]
The idea is that rendering at higher density makes a larger raster image from the vector file (which can be expanded without loss). So just resize to the desired size. One way I do that is to render at multiple densities of 72 then resize by the inverse amount (if you want to keep the nominal size).  So for example 4\*72 = 288. So I use -density 288. Then I resize by 1/4=25%. So -resize 25%.  You can do 16\*72=1152 density then resize by 1/16=6.25%

**Reply by hsweet3** [reply]
I tried this one and it seems to be getting better results... granted perception is part of the acceptance because I'm not sure there's a universal way to see or not see blur... Tried to get back to basics and not make it complex. Take a hig-res vector file and convert it to low-res. Low res, by definition is 72, 96 maybe... in this case we're sampling at 96, so density 300 in the output... doesn't meet the criteria

magick -quiet -colorspace sRGB [INPUT].eps[0] -set option:GETSIZE %wx%h +delete -density 2400 -units pixelsperinch [INPUT].eps[0] -resize %[GETSIZE] -transparent-color White -density 96 -sharpen 0x1 [OUTPUT].png

I'm not sure I can see it anymore, but 'they' say it's better

![FFO_0x1_2400_96](https://user-images.githubusercontent.com/22641367/205461244-0abf4a31-7d09-48fd-a78e-661998f1fa4c.png)

![IWT_0x1_2400_96](https://user-images.githubusercontent.com/22641367/205461250-982334e4-58a3-4606-87ba-0267f0a36b09.png)

**Reply by fmw42** [reply]
Your command reads the eps file twice. Is that really what you want to do? Seems like you only need to read it once.

**Reply by hsweet3** [reply]
I got that tip from @snibgo in order to obtain the GETSIZE info. Admittedly I don't really know what I'm doing so if the output is correct I don't care. If you think I can read it once to get the same result it'll save time executing the command

thanks

**Reply by codecrack3** [reply]
That's work for my issue. Thank you solution

**Comment by snibgo** [comment]
> Any help you can provide to get the actual inches of the file to match original will be much appreciated

I don't use Photoshop and can't comment on that. But if the image is 139 pixels wide and the desired physical dimension is 1.9267 inches, then you need a density of 139/1.9267 = 72.14 dots per inch.

> if I use the density 1200 to render it and then say -resize 25% will i consistently get 300DPI in the output? I just doing the simple math of 1200/4 = 300.

No. Try it and see. If an image is 1200 dpi, then resizing it by 25% won't change the desity. It remains 1200 dpi, and the physical dimension in inches will shrink. But if you change the density to 300, the physical dimension won't change much. (Probably not exactly the same, because the width in pixels may not be exactly divisible by 4.)

> Also - the output pngs seem to be about twice the file size.

Vector files and raster files store different types of data, so files sizes (in bytes) are not comparable. You might try a program such as pngcrush, or a different file format.

**Comment by fmw42** [comment]
your title says eps to jpg.  So why are you discussing png output?

**Reply by hsweet3** [reply]
initially we were trying to go to jpg but quickly discovered jpg doesn't support transparency so we're going with PNG vs gif - giff seems to have more ragged edges when using transparency

**Comment by fmw42** [comment]
PNG or ZIP or LZW compressed TIFF support transparency.

**Comment by snibgo** [comment]
Resizing, or any operation where output pixels are calculated from more than one input pixel, in linear RGB space is technically more accurate than in sRGB space. For ordinary photos there is (in my view) no practical difference. For cartoon images, where there are many "sudden" changes in colour, the difference is more obvious.

Another factor is the filter used for resize. The IM defaults work well for ordinary photos, but less well for cartoons. RoubidouxSharp or some other filter may be more acceptable.
```
magick -density 288 -colorspace sRGB IWT.eps[0] -colorspace RGB -filter robidouxsharp -resize 25% -colorspace sRGB x6.png
```
![x6](https://user-images.githubusercontent.com/33812368/205463255-7d95ba2c-c1e9-4ab9-952b-832c6c4b998b.png)

**Reply by hsweet3** [reply]
I'll have mostly illustrations but may have actual images also. I don't know if there's a method to identify a particular type of input image and then perform specific actions based on those inputs - perhaps as I get more familiar with IM it'll become more apparent
I will definitely test your suggestion above across the dozen or so images I'm using as my sample, thanks.
what is the 'x6' in above command?

**Comment by snibgo** [comment]
> what is the 'x6' in above command?

Huh? That is part of the output filename.
